### PR TITLE
support boxing when excuting loss.backward() in eager mode

### PR DESCRIPTION
**Pull request by wyushun**
根据这个Issue的讨论: [为nn.Parameter添加hook以提前执行更新参数所需的boxing过程](https://github.com/Oneflow-Inc/OneTeam/issues/1059), 做了下面的几个修改：
 - 在python的module.py中提供`global_param_grad_no_sync`接口
 - 在C++中定义一个全局状态变量，在pybind相关的接口代码中添加相应的get/set接口，用来操作这个状态
 - 在autograd_engine.cpp中的AccGrad4LeafTensor函数中添加具体的boxing处理

close oneflow-inc/oneteam#1059

### CODE REVIEW

**Comment by yuanms2** [comment]
有同事提议no_global_param_grad_sync 改成  global_param_grad_no_sync更好点

另外，这里提前做了boxing， 下游其它模块是不是就能自动知道不需要再做boxing了。

File: python/oneflow/nn/module.py [hunk 1]
```diff
@@ -66,6 +66,14 @@ def __init__(self):
     def forward(self, *args, **kwargs):
         raise NotImplementedError()
 
+    @contextmanager
+    def no_global_param_grad_sync(self):
+        flag = oneflow._oneflow_internal.GetGlobalParamGradSync()
+        if flag:
```
Comment (daquexian) [inline python/oneflow/nn/module.py]: 这个 if 是不是可以去掉

File: oneflow/core/autograd/autograd_engine.cpp [hunk 2]
```diff
@@ -25,6 +25,7 @@ limitations under the License.
 #include "oneflow/core/autograd/autograd_mode.h"
 #include "oneflow/core/eager/dev_vm_dep_object_consume_mode.h"
 #include "oneflow/core/functional/functional.h"
+#include "oneflow/api/python/env/env_api.h"
```
Comment (clackhan) [inline oneflow/core/autograd/autograd_engine.cpp:28]: ```api/python```中的代码不应该入侵到core目录下

File: oneflow/api/python/env/env.h [hunk 3]
```diff
@@ -93,6 +93,17 @@ inline Maybe<void> SetFLAGS_v(int32_t v_level) {
...
+inline Maybe<void> SetGlobalParamGradSync(bool flag) {
+  CHECK_NOTNULL_OR_RETURN((Global<bool, GlobalParamGradSync>::Get()));
+  *Global<bool, GlobalParamGradSync>::Get() = flag;
+  return Maybe<void>::Ok();
+}
+
+inline Maybe<bool> GetGlobalParamGradSync() {
+  auto flag = *Global<bool, GlobalParamGradSync>::Get();
+  return flag;
+}
```
Comment (clackhan) [inline oneflow/api/python/env/env.h:106]: 这两个接口的声明可以放在```oneflow/core/autograd/autograd_engine.h```，因为它们是autograd_engine相关的控制变量，而且只需导出Set接口，Get接口就放在```autograd_engine.cpp```中即可

File: oneflow/core/autograd/autograd_engine.cpp [hunk 4]
```diff
@@ -144,9 +145,28 @@ Maybe<void> FunctionNode::AccGrad4RetainGradTensor() {
...
+  // for (const std::shared_ptr<AutogradMeta>& out : output_meta_data_) {
+  for (auto i = 0; i < output_meta_data_.size(); i++) {
+    auto& out = output_meta_data_[i];
+
     if (out->is_leaf() && out->requires_grad()) {
       JUST(CopyOrAccGrad(out.get(), /*autograd_mode=*/false));
+
+      // control acc_grad to do boxing conditionally
+      const auto& acc_grad = out->acc_grad();
+      if (JUST(GetGlobalParamGradSync()) && acc_grad->is_consistent()) {
```
Comment (daquexian) [inline oneflow/core/autograd/autograd_engine.cpp:157]: CopyOrAccGrad 内部有一个遍历并执行 post_grad_accumulation_hooks() 的代码块，把这些逻辑（虽然不再是 hook）和那个代码块放在一起也许更好？

File: oneflow/api/python/env/env.cpp [hunk 5]
```diff
@@ -41,4 +41,6 @@ ONEFLOW_API_PYBIND11_MODULE("", m) {
   m.def("GetFLAGS_alsologtostderr", &GetFLAGS_alsologtostderr);
   m.def("SetFLAGS_v", &SetFLAGS_v);
   m.def("GetFLAGS_v", &GetFLAGS_v);
+  m.def("SetGlobalParamGradSync", &SetGlobalParamGradSync);
+  m.def("GetGlobalParamGradSync", &GetGlobalParamGradSync);
```
Comment (clackhan) [inline oneflow/api/python/env/env.cpp:45]: 这个接口仅控制auto grad行为，放在```oneflow/oneflow/api/python/autograd/autograd.cpp```即可

**Comment by wyushun** [comment]
> 有同事提议no_global_param_grad_sync 改成 global_param_grad_no_sync更好点

嗯，xiaoyu的这个名字建议更好一些，我更新一下

File: oneflow/core/autograd/autograd_engine.cpp (hunk 2 above)
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.cpp:28]: ok

File: oneflow/api/python/env/env.h (hunk 3 above)
Comment (wyushun) [reply oneflow/api/python/env/env.h:106]: 假设控制变量名字是flag，如果一开始flag是false，在python端的接口中如果没有导出get接口，会是下面这样：
```
def global_param_grad_no_sync(self):
    set_flag(false)
    yield
    set_flag(true)
```
这会导致最后的结果与一开始flag的值不一致，我感觉应该先把flag的值读出来，最后再set回去，你感觉这样有道理不 @clackhan

File: oneflow/api/python/env/env.cpp (hunk 5 above)
Comment (wyushun) [reply oneflow/api/python/env/env.cpp:45]: ok

File: oneflow/api/python/env/env.h (hunk 3 above)
Comment (clackhan) [reply oneflow/api/python/env/env.h:106]: 嗯，好的，那还是将Get导出

> 这会导致最后的结果与一开始flag的值不一致，我感觉应该先把flag的值读出来，最后再set回去，你感觉这样有道理不 @clackhan

嗯，好的，那还是将Get导出

File: oneflow/api/python/env/env.h [hunk 6]
```diff
@@ -93,6 +93,17 @@ inline Maybe<void> SetFLAGS_v(int32_t v_level) {
   return Maybe<void>::Ok();
 }
 inline Maybe<int32_t> GetFLAGS_v() { return FLAGS_v; }
+
+inline Maybe<void> SetGlobalParamGradSync(bool flag) {
```
Comment (daquexian) [inline oneflow/api/python/env/env.h:97]: 有没有办法通过测试（最好是可以每次都跑的单测）证明它的有效性呢：当它为 false 的时候，不发生任何的 boxing，而它为 true 的时候，只发生一次 boxing

File: python/oneflow/nn/module.py (hunk 1 above)
Comment (wyushun) [reply python/oneflow/nn/module.py]: 嗯，已经去掉了

File: oneflow/core/autograd/autograd_engine.cpp (hunk 4 above)
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.cpp:157]: 这部分逻辑本来是放在`CopyOrAccGrad `中的，昨天和yinggang讨论了下，他建议放到外面来，这部分还需要改不？

Comment (clackhan) [reply oneflow/core/autograd/autograd_engine.cpp:157]: > 这部分逻辑本来是放在`CopyOrAccGrad `中的，昨天和yinggang讨论了下，他建议放到外面来，这部分还需要改不？

原因也在这里说一下

File: oneflow/api/python/env/env.h (hunk 3 above)
Comment (clackhan) [reply oneflow/api/python/env/env.h:106]: > 现在的做法是先把初试的flag保存下来，最后再恢复回去，类似下面这样：
> 
> ```
> def global_param_grad_no_sync(self):
>     flag=get_flag()
>     set_flag(false)
>     yield
>     set_flag(flag)
> ```

好的，这样没问题

Comment (lixinqi) [reply oneflow/api/python/env/env.h:106]: 不要用`Global<bool, GlobalParamGradSync>`，应该用DECORATE(XXX, ThreadLocal)

Comment (wyushun) [reply oneflow/api/python/env/env.h:106]: ok

File: oneflow/core/autograd/autograd_engine.cpp (hunk 4 above)
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.cpp:157]: 参考上面yinggang的comment @clackhan

File: oneflow/core/autograd/autograd_engine.cpp [hunk 7]
```diff
@@ -144,9 +145,28 @@ Maybe<void> FunctionNode::AccGrad4RetainGradTensor() {
...
+        auto& tensor_info = output_tensor_infos_[i];
+        const auto& placement = JUST(tensor_info.GetPlacement());
+        const auto& nd_sbp = JUST(tensor_info.GetSBP());
+        std::vector<Symbol<SbpParallel>> sbp_tuple(nd_sbp->sbp_parallel().size());
+        for (auto i = 0; i < sbp_tuple.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.cpp:162]: 感谢，这里不该这么用

File: oneflow/core/autograd/autograd_engine.cpp [hunk 8]
```diff
@@ -144,9 +145,28 @@ Maybe<void> FunctionNode::AccGrad4RetainGradTensor() {
...
+    auto& out = output_meta_data_[i];
+
     if (out->is_leaf() && out->requires_grad()) {
       JUST(CopyOrAccGrad(out.get(), /*autograd_mode=*/false));
+
+      // control acc_grad to do boxing conditionally
+      const auto& acc_grad = out->acc_grad();
+      if (JUST(GetGlobalParamGradSync()) && acc_grad->is_consistent()) {
+        auto& tensor_info = output_tensor_infos_[i];
+        const auto& placement = JUST(tensor_info.GetPlacement());
```
[the remark this answers is not in the document]
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.cpp:159]: ok

File: oneflow/core/autograd/autograd_engine.cpp [hunk 9]
```diff
@@ -144,9 +145,28 @@ Maybe<void> FunctionNode::AccGrad4RetainGradTensor() {
...
+        const auto& nd_sbp = JUST(tensor_info.GetSBP());
+        std::vector<Symbol<SbpParallel>> sbp_tuple(nd_sbp->sbp_parallel().size());
+        for (auto i = 0; i < sbp_tuple.size(); ++i) {
+          sbp_tuple[i] = nd_sbp->sbp_parallel().Get(i);
+        }
```
[the remark this answers is not in the document]
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.cpp]: 已更新

File: oneflow/api/python/env/env.h [hunk 10]
```diff
@@ -93,6 +93,17 @@ inline Maybe<void> SetFLAGS_v(int32_t v_level) {
...
+  CHECK_NOTNULL_OR_RETURN((Global<bool, GlobalParamGradSync>::Get()));
+  *Global<bool, GlobalParamGradSync>::Get() = flag;
+  return Maybe<void>::Ok();
+}
+
+inline Maybe<bool> GetGlobalParamGradSync() {
+  auto flag = *Global<bool, GlobalParamGradSync>::Get();
```
[the remark this answers is not in the document]
Comment (wyushun) [reply oneflow/api/python/env/env.h]: 新奇在slack的comment直接让用`static thread_local bool`，已更新

**Comment by lixinqi** [comment]
> 有同事提议no_global_param_grad_sync 改成 global_param_grad_no_sync更好点

这个pr在开过程中会暂时搁置名字的争议，最终合并之前会考虑名字的决议。

> 另外，这里提前做了boxing， 下游其它模块是不是就能自动知道不需要再做boxing了。

会自动知道，因为下游optimizer要求输入的grad.sbp与param.sbp等同，而经过boxing之后这一条件就已满足，不再需要boxing。

**Comment by wyushun** [comment]
close oneflow-inc/oneteam#1059

File: oneflow/core/autograd/autograd_engine.h [hunk 11]
```diff
@@ -208,7 +208,11 @@ AutogradEngine* GetThreadLocalAutogradEngine();
 
 Maybe<void> AddAccumulateFunctionNode(const std::shared_ptr<Tensor>& tensor);
 
+Maybe<void> SetGlobalParamGradSync(bool flag);
+Maybe<bool> GetGlobalParamGradSync();
```
Comment (lixinqi) [inline oneflow/core/autograd/autograd_engine.h:212]: 这里的逻辑和AD没有关系，不应该放到autograd_engine.h里。

Comment (lixinqi) [reply oneflow/core/autograd/autograd_engine.h:212]: 单独写framework/global_param_grad_sync_guard.h。
不要提供Set接口，提供GlobalParamGardSyncGuard类。这个类还应该导出到python

File: python/oneflow/nn/module.py [hunk 12]
```diff
@@ -66,6 +67,13 @@ def __init__(self):
     def forward(self, *args, **kwargs):
         raise NotImplementedError()
 
+    @contextmanager
+    def global_param_grad_no_sync(self):
+        flag = oneflow._oneflow_internal.GetGlobalParamGradSync()
+        oneflow._oneflow_internal.SetGlobalParamGradSync(false)
+        yield
+        oneflow._oneflow_internal.SetGlobalParamGradSync(flag)
```
Comment (lixinqi) [inline python/oneflow/nn/module.py:75]:
```python
try:
    yield
finally:
    oneflow._oneflow_internal.SetGlobalParamGradSync(flag)
```
所有的contextmanager都应该写成这样。

Comment (lixinqi) [reply python/oneflow/nn/module.py:75]: 换成新版的GlobalParamGradSyncGuard之后，此处的代码应该写成：
```python
def global_param_grad_no_sync(self):
    guard = oneflow._oneflow_internal.GlobalParamGardSync(false)
    try:
        yield
    finally:
        del guard
```

File: oneflow/core/autograd/autograd_engine.cpp [hunk 13]
```diff
@@ -508,5 +523,14 @@ Maybe<void> AddAccumulateFunctionNode(const std::shared_ptr<Tensor>& tensor) {
   return Maybe<void>::Ok();
 }
 
+static thread_local bool global_param_grad_sync_flag = true;
```
[the remark this answers is not in the document]
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.cpp:526]: ok

File: oneflow/core/autograd/autograd_engine.cpp [hunk 14]
```diff
@@ -508,5 +523,14 @@ Maybe<void> AddAccumulateFunctionNode(const std::shared_ptr<Tensor>& tensor) {
...
+
+Maybe<void> SetGlobalParamGradSync(bool flag) {
+  global_param_grad_sync_flag = flag;
+  return Maybe<void>::Ok();
+}
```
[the remark this answers is not in the document]
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.cpp:531]: ok

File: oneflow/core/autograd/autograd_engine.h (hunk 11 above)
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.h:212]: ok

File: oneflow/core/framework/global_param_grad_sync_mode.h [hunk 15]
```diff
@@ -13,29 +13,29 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-#ifndef ONEFLOW_CORE_FRAMEWORK_GLOBAL_PARAM_GRAD_SYNC_GUARD_H_
-#define ONEFLOW_CORE_FRAMEWORK_GLOBAL_PARAM_GRAD_SYNC_GUARD_H_
+
+#ifndef ONEFLOW_CORE_FRAMEWORK_GLOBAL_PARAM_GRAD_SYNC_MODE_
+#define ONEFLOW_CORE_FRAMEWORK_GLOBAL_PARAM_GRAD_SYNC_MODE_
 
 namespace oneflow {
 
-class GlobalParamGardSyncGuard {
+struct GradSyncMode {
```
Comment (lixinqi) [inline oneflow/core/framework/global_param_grad_sync_mode.h:22]: GlobalGradSyncMode。加强语义。

File: oneflow/core/autograd/autograd_engine.cpp [hunk 16]
```diff
@@ -144,9 +146,23 @@ Maybe<void> FunctionNode::AccGrad4RetainGradTensor() {
...
+  for (auto i = 0; i < output_meta_data_.size(); i++) {
+    auto& out = output_meta_data_[i];
+
     if (out->is_leaf() && out->requires_grad()) {
       JUST(CopyOrAccGrad(out.get(), /*autograd_mode=*/false));
+
+      // control acc_grad to do boxing conditionally
+      const auto& acc_grad = out->acc_grad();
+      if (GradSyncMode::is_enabled() && acc_grad->is_consistent()) {
+        auto& tensor_info = output_tensor_infos_[i];
```
[the remark this answers is not in the document]
Comment (wyushun) [reply oneflow/core/autograd/autograd_engine.cpp:159]: 不赞同，i已经限定了作用域，用.at(i)没有更多的益处，下面是标准库中vector.at的大概实现，主要是做了下标检查：
```
      reference at(size_type __n)
      {
	if (__n >= this->size())
	  __throw_out_of_range_fmt(__N("vector::_M_range_check: __n "
				       "(which is %zu) >= this->size() "
				       "(which is %zu)"),
				   __n, this->size());
	return (*this)[__n];
      }
```
operator[]则不会做这个检查：
```
      reference operator[](size_type __n) _GLIBCXX_NOEXCEPT
      {
	__glibcxx_requires_subscript(__n);
	return *(this->_M_impl._M_start + __n);
      }
```

如果在已经限定了下标访问作用域的话，我更倾向用operator[]，没必要在.at中花额外的开销来做范围检查，这个开销看起来很小，但凡是不能被编译器优化掉的if，都可能会导致cpu执行过程中的分支预测开销，也许编译器可以根据上下文把这个if优化掉，但上下文复杂的话也许优化不掉，相比之下，operator[]肯定不会有这个问题。

**Comment by hjchen2** [comment]
static zero的部分工作是不是可以单独提一个PR，这个PR可以先合并

File: oneflow/core/autograd/autograd_meta.h [hunk 17]
```diff
@@ -103,6 +103,8 @@ class TensorInfo final {
   explicit TensorInfo(const Tensor& tensor);
 
   Maybe<Tensor> zeros() const;
+  Optional<Symbol<ParallelDesc>> placement() { return parallel_desc_; }
+  Optional<Symbol<NdSbp>> sbp() { return nd_sbp_; }
```
Comment (hjchen2) [inline oneflow/core/autograd/autograd_meta.h:107]:
```suggestion
  Optional<Symbol<NdSbp>> sbp() const { return nd_sbp_; }
```

File: oneflow/core/autograd/autograd_meta.h [hunk 18]
```diff
@@ -103,6 +103,8 @@ class TensorInfo final {
   explicit TensorInfo(const Tensor& tensor);
 
   Maybe<Tensor> zeros() const;
+  Optional<Symbol<ParallelDesc>> placement() { return parallel_desc_; }
```
Comment (hjchen2) [inline oneflow/core/autograd/autograd_meta.h:106]:
```suggestion
  Optional<Symbol<ParallelDesc>> placement() const { return parallel_desc_; }
```

**Review by hjchen2 — APPROVED** [review]

**Review by clackhan — APPROVED** [review]
